learn_evm/arithmetic-checks.md

```diff
@@ -0,0 +1,668 @@
+            overflow := xor(slt(a, 0), sgt(b, c))
+        }
+
+        if (overflow) arithmeticError();
+    }
+}
+```
+
+A different approach to detecting overflow in addition is to observe that adding two integers with different signs will never overflow.
+This reduces the check to the case when both summands have the same sign.
+If the sign of the sum is different from one of the summands, the result has overflowed.
+
+```solidity
+function checkedAddInt4(int256 a, int256 b) public pure returns (int256 c) {
+    unchecked {
+        c = a + b;
+
+        // Overflow, if the signs of `a` and `b` are the same,
+        // but the sign of the result `c = a + b` differs from its summands.
+        // When the signs of `a` and `b` differ overflow is not possible.
+        if ((~a ^ b) & (a ^ c) < 0) arithmeticError();
+    }
+}
+```
+
+Rather than checking the sign bit explicitly, which can be achieved by shifting the value to the right by 255 bits and checking that it is non-zero,
+we can use the `slt` operation to compare the value with `0`.
+
+## Arithmetic checks for uint256 subtraction
+
+The process of checking for underflow in subtraction is akin to that of addition.
+If we subtract `a - b`, and `b` is greater than `a`, we have an underflow situation.
+
+```solidity
+function checkedSubUint(uint256 a, uint256 b) public pure returns (uint256 c) {
+    unchecked {
+        c = a - b;
+
+        if (b > a) arithmeticError();
+    }
+}
+```
+
+We could, like before, perform the check on the result itself using `if (c > a) arithmeticError();`, because subtracting a positive value from `a` should yield a value less than or equal to `a`.
+However, in this case, we don't save any operations.
+
+Just as with addition, for signed integers, we can combine the checks for both scenarios into a single check using `xor`.
+
+```solidity
+function checkedSubInt(int256 a, int256 b) public pure returns (int256 c) {
+    unchecked {
+        c = a - b;
+
+        bool overflow;
+
+        assembly {
+            // If `b >= 0`, then the result `c = a - b` can't be greater than `a`.
+            // If `b <  0`, then the result `c = a - b` can't be less than `a`.
+            overflow := xor(sgt(b, 0), sgt(a, c))
+        }
+
+        if (overflow) arithmeticError();
+    }
+}
+```
+
+## Arithmetic checks for uint256 multiplication
+
+To detect overflow when multiplying two unsigned integers, we can again use the approach of computing the maximum possible value of a multiplicand and check that it isn't exceeded.
+
+```solidity
+/// @notice versions >=0.8.0 && <0.8.17
+function checkedMulUint1(uint256 a, uint256 b) public pure returns (uint256 c) {
+    unchecked {
+        c = a * b;
+
+        if (a != 0 && b > type(uint256).max / a) arithmeticError();
+    }
+}
+```
+
+> It's important to note that the Solidity compiler always includes a division by zero check for all division and modulo operations, regardless of the presence of an unchecked block.
```

Comment:
division by zero and modulo by zero check maybe? saying "a division by zero" check on a modulo is a bit confusing ;) (even if its not false)